cmd/agent/app/throttling/credit_store.go

```diff
@@ -0,0 +1,261 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package throttling
+
+import (
+	"io"
+	"math"
+	"sync"
+	"time"
+)
+
+// tokenBucket is a simple token bucket used to refill credits for services
+// with increasing time.
+type tokenBucket struct {
+	creditsPerSecond float64
+	balance          float64
+	maxBalance       float64
+	lastTick         time.Time
+	timeNow          func() time.Time // For testing
+}
+
+func newTokenBucket(options CreditAccruerOptions, timeNow func() time.Time) *tokenBucket {
+	return &tokenBucket{
+		creditsPerSecond: options.CreditsPerSecond,
+		balance:          options.MaxBalance,
+		maxBalance:       options.MaxBalance,
+		timeNow:          timeNow,
+		lastTick:         timeNow(),
+	}
+}
+
+// Withdraw deducts as much of the total balance possible without exceeding
+// maxWithdrawal.
+func (t *tokenBucket) Withdraw(maxWithdrawal float64) float64 {
+	now := t.timeNow()
+	interval := now.Sub(t.lastTick)
+	t.lastTick = now
+	diff := interval.Seconds() * t.creditsPerSecond
+	t.balance = math.Min(t.balance+diff, t.maxBalance)
+	result := math.Min(t.balance, maxWithdrawal)
+	t.balance -= result
+	return result
+}
+
+// client represents a Jaeger-instrumented caller. Every Jaeger client
+// corresponds to an instrumented service, but it may be one of many instances
+// of that service. Therefore, there is a one-to-many relationship between
+// services and clients. The clients must request credits from the service's
+// creditAccruer and they all share the same underlying tokenBucket.
+// For example, if one client requests a withdrawal and receives all the service
+// credits, the next client withdrawal request will be rejected until the
+// credits refill.
+type client struct {
+	// id is a unique id for the instance of the Jaeger client
+	id string
+	// perOperationBalance maintains per-operation balances of withdrawn credits
+	perOperationBalance map[string]float64
+	// updateTime is the last time client was updated
+	updateTime time.Time
+}
+
+func newClient(id string, currentTime time.Time) *client {
+	return &client{
+		id:                  id,
+		updateTime:          currentTime,
+		perOperationBalance: map[string]float64{},
+	}
+}
+
+// Spend depletes the balance from the client for the given operation. The usage
+// would be when a client submits spans for operation X, it is spending those
+// credits it requested for operation X.
+func (c *client) Spend(operationName string, credits float64) bool {
+	balance := c.perOperationBalance[operationName]
+	if credits > balance {
+		return false
+	}
+
+	balance -= credits
+	c.perOperationBalance[operationName] = balance
+	return true
+}
+
+// CreditAccruerOptions provides values to be used with a creditAccruer object
+type CreditAccruerOptions struct {
+	MaxOperations    int
+	CreditsPerSecond float64
+	MaxBalance       float64
+	ClientMaxBalance float64
```

Comment:
- what is the difference between these two max-balance values?
- we will vary likely need different limits and refill rates per service, is this design extensible?

Reply:
I'll add this to the documentation, but `MaxBalance` describes the maximum for the token bucket, and `ClientMaxBalance` refers to the maximum a client can attain before it must spend the credits (`Withdraw` returns `0` for these clients).

The reason I use a `struct` instead of using constants was to allow for extensibility. In the future, we can use a default for most services, and add overrides for specific services that use custom values.